parsec/backend/app.py

```diff
@@ -233,16 +240,68 @@ def __init__(self, config, event_bus=None):
         return {"status": "ok", "pong": msg["ping"]}
 
     async def _api_blockstore_post(self, client_ctx, msg):
-        if not self.blockstore:
+        async def blockstore_post(send_channel, func, *args, **kwargs):
+            async with send_channel:
+                try:
+                    result = await func(*args, **kwargs)
+                except ParsecError as exc:
+                    result = exc
+                await send_channel.send(result)
+
+        if not self.blockstores:
             return {"status": "not_available", "reason": "Blockstore not available"}
 
-        return await self.blockstore.api_blockstore_post(client_ctx, msg)
+        async with trio.open_nursery() as nursery:
+            send_channel, receive_channel = trio.open_memory_channel(0)
+            async with send_channel:
+                for blockstore in self.blockstores:
+                    nursery.start_soon(
+                        blockstore_post,
+                        send_channel.clone(),
+                        blockstore.api_blockstore_post,
+                        client_ctx,
+                        msg,
+                    )
+
+            values = [value async for value in receive_channel]
+        values = sorted(values, key=lambda x: not isinstance(x, ParsecError))
+        if isinstance(values[0], ParsecError):
+            raise values[0]
+        return values[0]  # Assume that all responses are the same
 
     async def _api_blockstore_get(self, client_ctx, msg):
-        if not self.blockstore:
+        async def blockstore_get(send_channel, func, *args, **kwargs):
+            async with send_channel:
+                try:
+                    result = await func(*args, **kwargs)
+                except ParsecError as exc:
+                    result = exc
+                await send_channel.send(result)
+
+        if not self.blockstores:
             return {"status": "not_available", "reason": "Blockstore not available"}
 
-        return await self.blockstore.api_blockstore_get(client_ctx, msg)
+        async with trio.open_nursery() as nursery:
+            send_channel, receive_channel = trio.open_memory_channel(0)
+            async with send_channel:
+                for blockstore in self.blockstores:
+                    nursery.start_soon(
+                        blockstore_get,
+                        send_channel.clone(),
+                        blockstore.api_blockstore_get,
+                        client_ctx,
+                        msg,
+                    )
+
+            values = []
+            async for value in receive_channel:
+                values.append(value)
+                if not isinstance(value, ParsecError):
+                    nursery.cancel_scope.cancel()
+        values = sorted(values, key=lambda x: isinstance(x, ParsecError))
+        if isinstance(values[0], ParsecError):
```

Comment:
Tu cherchais à faire un `next((x for x in vales if isinstance(x, ParsecError)), False)` ?

Reply:
OK